File: days/day14.swift
#!/usr/bin/env swift
import Foundation


struct V2 : Hashable {
    var x: Int
    var y: Int
}


func part1(_ data: String, N: Int = 100, w: Int = 101, h: Int = 103) -> Int {
    let data = data.matches(of: #/[+-]?\d+/#).map { Int($0.0)! }

    var grid: [V2:Int] = [:]
    for i in stride(from: 0, to: data.count, by: 4) {
        let px = data[i+0]
        let py = data[i+1]
        let vx = data[i+2]
        let vy = data[i+3]
        let x = (px + vx * N) % w
        let y = (py + vy * N) % h
        let p = V2(x: (x + w) % w, y: (y + h) % h)
        grid[p, default: 0] += 1
    }

    let w2 = w / 2
    let h2 = h / 2
    let w2p = w2 + 1
    let h2p = h2 + 1
    var qs: [V2:Int] = [:]

    for (p, n) in grid {
        if p.x == w2 || p.y == h2 {
            continue
        }
        let q = V2(x: p.x / w2p, y: p.y / h2p)
        qs[q, default: 0] += n
    }

    var ans = 1
    for n in qs.values {
        ans *= n
    }
    return ans
}


func part2(_ data: String, w: Int = 101, h: Int = 103) -> Int {
    let data = data.matches(of: #/[+-]?\d+/#).map { Int($0.0)! }

    for t in 1... {
        var grid: [V2:Int] = [:]
        for i in stride(from: 0, to: data.count, by: 4) {
            let px = data[i+0]
            let py = data[i+1]
            let vx = data[i+2]
            let vy = data[i+3]
            let x = (px + vx * t) % w
            let y = (py + vy * t) % h
            let p = V2(x: (x + w) % w, y: (y + h) % h)
            grid[p, default: 0] += 1
        }

        if (grid.values.allSatisfy { $0 == 1 }) {
            for p in grid.keys {
                if ((0..<20).allSatisfy { grid[V2(x: p.x + $0, y: p.y)] != nil }) {
                    return t
                }
            }
        }
    }
    fatalError()
}


let test = """
p=0,4 v=3,-3
p=6,3 v=-1,-3
p=10,3 v=-1,2
p=2,0 v=2,-1
p=0,0 v=1,3
p=3,0 v=-2,-2
p=7,6 v=-1,-3
p=3,0 v=-1,-2
p=9,3 v=2,3
p=7,3 v=-1,2
p=2,4 v=2,-3
p=9,5 v=-3,-3
"""
assert(part1(test, w: 11, h: 7) == 12)


let data = try! String(contentsOfFile: "day14.in", encoding: .utf8)
let ans1 = part1(data)
print("part1: \(ans1)")
let ans2 = part2(data)
print("part2: \(ans2)")

File: days/day14.py
#!/usr/bin/env python
import io
import itertools
import re
from collections import Counter


def part1(data, N=100, w=101, h=103):
    data = list(map(int, re.findall(r'\-?\d+', data)))
    px = data[0::4]
    py = data[1::4]
    vx = data[2::4]
    vy = data[3::4]

    grid = Counter()
    for i in range(len(px)):
        x = (px[i] + vx[i] * N) % w
        y = (py[i] + vy[i] * N) % h
        grid[x,y] += 1

    w2 = w // 2
    h2 = h // 2
    qs = Counter()

    for (x,y), n in grid.items():
        if x == w2 or y == h2:
            continue
        qs[x // (w2+1), y // (h2+1)] += n

    ans = 1
    for n in qs.values():
        ans *= n
    return ans


def part2(data, w=101, h=103):
    data = list(map(int, re.findall(r'\-?\d+', data)))
    px = data[0::4]
    py = data[1::4]
    vx = data[2::4]
    vy = data[3::4]

    for t in itertools.count(1):
        grid = Counter()
        for i in range(len(px)):
            x = (px[i] + vx[i] * t) % w
            y = (py[i] + vy[i] * t) % h
            grid[x,y] += 1

        if all(n == 1 for n in grid.values()):
            for ox,oy in grid:
                if all(grid.get((ox+i, oy)) for i in range(20)):
                    return t


def display(grid, w, h):
    with io.StringIO() as so:
        for y in range(h):
            for x in range(w):
                c = grid.get((x,y), '.')
                print(c, end='', file=so)
            print(file=so)
        print(so.getvalue())


data = '''
p=0,4 v=3,-3
p=6,3 v=-1,-3
p=10,3 v=-1,2
p=2,0 v=2,-1
p=0,0 v=1,3
p=3,0 v=-2,-2
p=7,6 v=-1,-3
p=3,0 v=-1,-2
p=9,3 v=2,3
p=7,3 v=-1,2
p=2,4 v=2,-3
p=9,5 v=-3,-3
'''
assert part1(data, w=11, h=7) == 12


data = open('day14.in').read()
ans = part1(data)
print('part1:', ans)
ans = part2(data)
print('part2:', ans)
